### PR TITLE
Add support to filter files based on their age

### PR DESCRIPTION
**Pull request by meirelles**
It's very useful to my usage. I'm use to move very old files to cheaper alternatives. I hope be useful to others.

Thank you

### CODE REVIEW

**Comment by ncw** [comment]
Great idea - thanks.

Can you
  * make sure the builds pass (hint - you need to `go fmt` your code)
  * add tests (in `filter_test.go`) 

I'm not particulary happy with the unit of seconds for the parameters.  How about building an option parser based off `time.Duration` and [time.ParseDuration]( https://golang.org/pkg/time/#ParseDuration)- see [SizeSuffix](https://github.com/ncw/rclone/blob/master/fs/config.go#L67) for an example.  That would let the user specify `--max-age 24h` or `--min-age 32s`.  Have a go with this if you like, or leave it for the moment and I'll do it after this lands.

When you are done can you [squash your branch and force push](http://stackoverflow.com/questions/14534397/squash-all-my-commits-into-one-for-github-pull-request).

Thanks

Nick

**Comment by meirelles** [comment]
I'll do! Thanks for the suggestions. I never wrote a single line of Go before submit this patch.

**Comment by ncw** [comment]
You are doing well so far!  I suggest you ignore my suggestion for the Duration parser for the moment and we'll work on just getting this merged first.

**Comment by meirelles** [comment]
It's good?

File: fs/filter.go [hunk 1]
```diff
@@ -23,6 +24,8 @@ var (
 	includeRule    = pflag.StringP("include", "", "", "Include files matching pattern")
 	includeFrom    = pflag.StringP("include-from", "", "", "Read include patterns from file")
 	filesFrom      = pflag.StringP("files-from", "", "", "Read list of source-file names from file")
+	minAge         = pflag.Int64P("min-age", "", 0, "Don't transfer any file younger than N seconds")
```
Comment (klauspost) [inline fs/filter.go]: I would suggest you use a `time.Duration`, see [DurationP](https://godoc.org/github.com/spf13/pflag#DurationP). This makes it much easier to specify minutes, hours, etc.

**Comment by klauspost** [comment]
It would also help a lot, if you included [documentation update](https://github.com/ncw/rclone/blob/master/docs/content/filtering.md) for the new option.

**Comment by ncw** [comment]
I couple of other things I thought of
   * we should avoid calling o.ModTime() unless it is necessary as it can be an expensive operation
   * there is a perfectly good time.Duration parser for the command line flags so we should use that!

**Comment by meirelles** [comment]
@klauspost my English is pretty bad. I'm not confident enough to write documentation. Somebody else should take the documentation update.

I utilized the DurationP at first version. But later I changed my mind as in practice I would use days, months or years to express age of files. Also I found a bit misleading be limited to hours. I did let this opened because the best solution would be write a custom parser which works with all that stuff IMHO, which defaults to seconds if no suffix is provided at all. Make sense?

I totally agree with o.ModTime(). It was my mistake assume it has been cached somewhere during the fetching of listing.

**Comment by ncw** [comment]
> I'm not confident enough to write documentation. Somebody else should take the documentation update.

I'm happy to do that if you aren't confident with it.

> I utilized the DurationP at first version. But later I changed my mind as in practice I would use days, months or years to express age of files. Also I found a bit misleading be limited to hours. I did let this opened because the best solution would be write a custom parser which works with all that stuff IMHO, which defaults to seconds if no suffix is provided at all. Make sense?

I think I'd like to make sure we have a proper parser before I make a release with this in, but I'm happy to run with seconds or a DurationP for the moment.

I'd quite like to extend the Duration parser so it understands days `d`, weeks `w` and years `y` too.  There is a reason that the go team haven't done this already and that is that all of those units aren't necessarily a standard length.  However for rclone I'm not so worried about that!

> I totally agree with o.ModTime(). It was my mistake assume it has been cached somewhere during the fetching of listing.

It probably has been cached, but not definitely depending on exactly what the user is doing.

**Comment by meirelles** [comment]
@ncw, why the o.Size() is cached and o.ModTime() isn't? They are both very important to check and sync. Would be a performance gain for operations.go keep it cached, normally both are returned together at the listing APIs. Perhaps only updating/invalidating after the Update/Remove call.

I'm quite busy. If you give few days/one week I'll made a custom parser to complete this feature in the right way, sounds good to you?

**Comment by ncw** [comment]
@meirelles take your time - that is fine!

I'll review the commit in detail at the weekend.

I'm planning to make a release at the weekend, but I'll leave this one out of it for the moment as I've got some urgent fixes I need to push out.

You also asked (but I think you may have deleted it now?)
> why the o.Size() is cached and o.ModTime() isn't? They are both very important to check and sync. Would be a performance gain for operations.go keep it cached, normally both are returned together at the listing APIs. Perhaps only updating/invalidating after the Update/Remove call.

We cache it where we can, but not all remotes return it in the directory listing.  Drive/google compute storage/onedrive do but the others don't.  Otherwise it works as you suggest.

**Comment by meirelles** [comment]
Let's check it

**Comment by ncw** [comment]
I've reviewed your commit - looking really good thanks.

I've added a few minor points to the code - can you fixup, rebase and force push please?

I've also fixed the build so can you make sure the build for this branch goes green too.

Thanks, and apologies for the delay

Nick

PS I'll write the docs after I merge it.

**Comment by meirelles** [comment]
Great. I'll do my best to comment after we decide how should behave the parseAge function. Thanks.

**Comment by ncw** [comment]
Thanks for the input.  I think we should get this merged as soon as possible, before the master branch diverges too much!

Cheers

Nick

**Comment by meirelles** [comment]
Commented. Resync'ed with master. Let's merge it. Thanks.

**Comment by ncw** [comment]
I've merged this.  I've also committed a fix! e69e1810901727f0078adedcef815aa44948103f

Thank you very much for your contribution :-)

**Comment by ei-grad** [comment]
:+1:

**Comment by meirelles** [comment]
Looks great! Thank you.

**Pull request merged**